File: greythr_automation.js
const { chromium } = require('playwright');
const fs = require("fs");

// Get credentials from environment variables (GitHub Secrets)
const getConfig = () => {
  return {
    url: process.env.GREYTHR_URL || "https://kalvium.greythr.com/",
    username: process.env.LOGIN_ID,
    password: process.env.LOGIN_PASSWORD,
    location: process.env.SIGNIN_LOCATION || "Office",
    debugMode: process.env.DEBUG_MODE === 'true',
    isManualRun: process.env.MANUAL_RUN === 'true'
  };
};

const automate = async () => {
  const config = getConfig();
  
  // Validate credentials
  if (!config.username || !config.password) {
    console.error("ERROR: LOGIN_ID and LOGIN_PASSWORD environment variables are required");
    process.exit(1);
  }

  console.log("Starting GreytHR automation...");
  console.log(`URL: ${config.url}`);
  console.log(`Debug Mode: ${config.debugMode}`);
  console.log(`Manual Run: ${config.isManualRun}`);

  // Launch browser with more robust settings
  const browser = await chromium.launch({
    headless: !config.debugMode, // Headless in production, visible in debug
    slowMo: 250, // Delay between actions for stability
    args: [
      '--no-sandbox',
      '--disable-setuid-sandbox',
      '--disable-dev-shm-usage',
      '--disable-accelerated-2d-canvas',
      '--disable-gpu',
      '--window-size=1920,1080'
    ]
  });
  
  const context = await browser.newContext({
    viewport: { width: 1920, height: 1080 },
    userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
  });
  
  const page = await context.newPage();
  
  // Add console listener for debugging
  if (config.debugMode) {
    page.on('console', msg => console.log('Browser console:', msg.text()));
  }
  
  // Set longer timeout
  page.setDefaultTimeout(60000);
  
  var result = "";
  
  try {
    console.log("Navigating to login page...");
    await page.goto(config.url, { 
      waitUntil: 'domcontentloaded',
      timeout: 60000 
    });
    
    // Wait for page to be interactive
    await page.waitForLoadState('networkidle');
    console.log("Page loaded successfully");
    
    // Save screenshot if in debug mode
    if (config.debugMode) {
      await page.screenshot({ path: "/tmp/initial_page.png" });
    }
    
    // Wait for and fill username
    console.log("Looking for username field...");
    const usernameField = await page.waitForSelector('input[type="text"], input[type="email"], input[name="username"]', { timeout: 30000 });
    await usernameField.fill(config.username);
    console.log("Username entered");
    
    // Wait for and fill password
    console.log("Looking for password field...");
    const passwordField = await page.waitForSelector('input[type="password"]', { timeout: 10000 });
    await passwordField.fill(config.password);
    console.log("Password entered");
    
    // Submit form
    console.log("Submitting login form...");
    await Promise.race([
      page.click('button[type="submit"]'),
      page.click('input[type="submit"]'),
      page.press('input[type="password"]', 'Enter')
    ]);
    
    // Wait for navigation after login
    console.log("Waiting for login to complete...");
    await page.waitForLoadState('networkidle', { timeout: 30000 });
    
    console.log("Current URL:", page.url());
    
    if (config.debugMode) {
      await page.screenshot({ path: "/tmp/dashboard.png" });
    }
    
    // Check if already signed in (Sign Out button exists)
    const signOutExists = await page.locator('button:has-text("Sign Out")').count();
    if (signOutExists > 0) {
      console.log("Already signed in (Sign Out button found)");
      result = "✅ Already signed in - no action needed";
      if (config.debugMode) {
        await page.screenshot({ path: "/tmp/already_signed_in.png" });
      }
      await browser.close();
      return result;
    }
    
    console.log("Not signed in yet, proceeding...");
    
    // Look for the Sign In button in the attendance section
    console.log("Looking for Sign In button in attendance section...");
    
    // Wait for and click the Sign In button
    const signInButton = await page.locator('button:has-text("Sign In")').first();
    await signInButton.waitFor({ state: 'visible', timeout: 10000 });
    console.log("Found Sign In button");
    
    await signInButton.click();
    console.log("Clicked initial Sign In button");
    
    // Wait for modal with a longer timeout
    await page.waitForTimeout(4000);
    console.log("Waiting for modal to appear...");
    
    // Save debug info
    if (config.debugMode) {
      await page.screenshot({ path: "/tmp/modal_state.png" });
      const htmlContent = await page.content();
      fs.writeFileSync('/tmp/modal_page.html', htmlContent);
      console.log("Saved debug files to /tmp/");
    }
    
    // Check what's visible on the page
    const debugInfo = await page.evaluate(() => {
      const info = {
        modals: [],
        dropdowns: [],
        buttons: [],
        modalPresent: false
      };
      
      // Check for modals
      const modals = document.querySelectorAll('gt-popup-modal');
      modals.forEach((modal, index) => {
        info.modals.push({
          index,
          hasOpen: modal.hasAttribute('open'),
          isVisible: modal.offsetParent !== null
        });
      });
      
      // Check for dropdowns
      const dropdowns = document.querySelectorAll('gt-dropdown, .dropdown-button, button.dropdown-button, [class*="dropdown"]');
      dropdowns.forEach((dd, index) => {
        if (dd.offsetParent !== null) {
          info.dropdowns.push({
            index,
            tagName: dd.tagName,
            className: dd.className,
            text: dd.textContent?.trim().substring(0, 50),
            isVisible: true
          });
        }
      });
      
      // Check for all visible buttons
      const buttons = document.querySelectorAll('button');
      buttons.forEach((btn) => {
        if (btn.offsetParent !== null && btn.textContent.trim()) {
          info.buttons.push(btn.textContent.trim());
        }
      });
      
      // Check if modal text is present
      const modalTexts = ['Tell us your work location', 'You are not signed in yet', 'Sign-In Location'];
      for (const text of modalTexts) {
        const elements = Array.from(document.querySelectorAll('*')).filter(el => 
          el.textContent.includes(text) && el.offsetParent !== null
        );
        if (elements.length > 0) {
          info.modalPresent = true;
          break;
        }
      }
      
      return info;
    });
    
    console.log("Debug Info:", JSON.stringify(debugInfo, null, 2));
    
    // If modal is present, handle it
    if (debugInfo.modalPresent || debugInfo.modals.length > 0) {
      console.log("Modal detected, attempting to handle...");
      
      // Method 1: Try using Playwright's built-in selectors
      try {
        // Look for dropdown with Office text
        const dropdownWithOffice = await page.locator('button:has-text("' + config.location + '")').first();
        if (await dropdownWithOffice.count() > 0) {
          console.log(`Found dropdown with ${config.location} text, clicking...`);
          await dropdownWithOffice.click();
          await page.waitForTimeout(2000);
          
          // Look for Office option in expanded dropdown
          const officeOptions = await page.locator('div:has-text("' + config.location + '")').all();
          for (const option of officeOptions) {
            const text = await option.textContent();
            if (text?.trim() === config.location) {
              console.log(`Clicking ${config.location} option...`);
              await option.click();
              break;
            }
          }
          
          await page.waitForTimeout(2000);
          
          // Click Sign In button
          const modalSignInButtons = await page.locator('button:has-text("Sign In")').all();
          if (modalSignInButtons.length > 1) {
            console.log(`Found ${modalSignInButtons.length} Sign In buttons, clicking the last one...`);
            await modalSignInButtons[modalSignInButtons.length - 1].click();
          } else if (modalSignInButtons.length === 1) {
            console.log("Clicking the only Sign In button...");
            await modalSignInButtons[0].click();
          }
          
          result = "✅ Modal handled with Playwright selectors";
        }
      } catch (e) {
        console.log("Playwright selector method failed:", e.message);
      }
      
      // Method 2: Direct DOM manipulation if Method 1 failed
      if (!result.includes("✅")) {
        console.log("Trying direct DOM manipulation...");
        
        const manipResult = await page.evaluate((location) => {
          try {
            // Find and click any button containing location text
            const locationButtons = Array.from(document.querySelectorAll('button')).filter(b => 
              b.textContent.includes(location) && b.offsetParent !== null
            );
            
            if (locationButtons.length > 0) {
              locationButtons[0].click();
              console.log(`Clicked ${location} button`);
              
              // Wait a bit and try to find location option
              setTimeout(() => {
                const allDivs = Array.from(document.querySelectorAll('div'));
                const locationDiv = allDivs.find(d => 
                  d.textContent.trim() === location && 
                  !d.querySelector('button') // Not the button container
                );
                if (locationDiv) {
                  locationDiv.click();
                  console.log(`Clicked ${location} div`);
                }
                
                // Then click Sign In
                setTimeout(() => {
                  const signInButtons = Array.from(document.querySelectorAll('button')).filter(b => 
                    b.textContent.trim() === 'Sign In' && b.offsetParent !== null
                  );
                  if (signInButtons.length > 0) {
                    signInButtons[signInButtons.length - 1].click();
                    console.log("Clicked Sign In");
                  }
                }, 1000);
              }, 1000);
              
              return "Attempted DOM manipulation";
            }
            
            return `No ${location} button found`;
          } catch (error) {
            return error.message;
          }
        }, config.location);
        
        console.log("DOM manipulation result:", manipResult);
        await page.waitForTimeout(4000);
      }
      
    } else {
      console.log("No modal detected on page");
      result = "❌ Modal did not appear after clicking Sign In";
    }
    
    // Final verification
    await page.waitForTimeout(3000);
    console.log("Performing final verification...");
    
    // Check for success indicators
    const signOutFinal = await page.locator('button:has-text("Sign Out")').count();
    const successMessage = await page.locator('text=/successfully/i').count();
    
    if (signOutFinal > 0) {
      result = "✅ Successfully signed in - Sign Out button present";
      console.log("SUCCESS: Sign Out button found");
    } else if (successMessage > 0) {
      result = "✅ Successfully signed in - Success message present";
      console.log("SUCCESS: Success message found");
    } else if (!result.includes("✅")) {
      result = "⚠️ Sign in attempted but could not verify success";
      console.log("WARNING: Could not verify sign in success");
    }
    
    // Take final screenshot
    if (config.debugMode) {
      await page.screenshot({ path: "/tmp/final_result.png" });
      const finalHtml = await page.content();
      fs.writeFileSync('/tmp/final_page_source.html', finalHtml);
    }
    
  } catch (error) {
    console.log("Error during automation:", error.message);
    if (config.debugMode) {
      await page.screenshot({ path: "/tmp/error_screenshot.png" });
    }
    result = "❌ Error: " + error.message;
  }
  
  await browser.close();
  return result;
};

const signIn = async () => {
  const date = new Date();
  let result = await automate();
  
  // Log result
  const logMessage = `${result} at ${date.getHours()}:${date.getMinutes()} on ${date.getDate()}/${date.getMonth() + 1}/${date.getFullYear()}`;
  
  console.log("\n" + "=".repeat(50));
  console.log("FINAL RESULT:", result);
  console.log("Time:", new Date().toLocaleString());
  console.log("=".repeat(50));
  
  // Append to log file if exists
  try {
    fs.appendFileSync("log.txt", `\n${logMessage}`);
  } catch (e) {
    // Log file may not exist in CI environment
    console.log("Could not write to log file:", e.message);
  }
  
  // Exit with appropriate code for GitHub Actions
  if (result.includes("✅")) {
    process.exit(0); // Success
  } else if (result.includes("⚠️")) {
    console.log("Warning: Sign-in status unclear");
    process.exit(0); // Don't fail the workflow for warnings
  } else {
    process.exit(1); // Failure
  }
};

// Run sign-in immediately when script is executed
signIn();

File: greythr_automation.py
#!/usr/bin/env python3
"""
GreytHR Auto Sign-In Automation Script
Compatible with GitHub Actions workflow
"""

import os
import sys
import time
import logging
from datetime import datetime
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from selenium.common.exceptions import TimeoutException, NoSuchElementException
import json

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.StreamHandler(sys.stdout)
    ]
)
logger = logging.getLogger(__name__)

class GreytHRAutomation:
    def __init__(self):
        # Get environment variables
        self.login_url = os.environ.get('LOGIN_URL', 'https://kalvium.greythr.com/')
        self.login_id = os.environ.get('LOGIN_ID', '')
        self.login_password = os.environ.get('LOGIN_PASSWORD', '')
        self.signin_location = os.environ.get('SIGNIN_LOCATION', 'Office')
        self.debug_mode = os.environ.get('DEBUG_MODE', 'false').lower() == 'true'
        self.manual_run = os.environ.get('MANUAL_RUN', 'false').lower() == 'true'
        
        # Validate credentials
        if not self.login_id or not self.login_password:
            logger.error("LOGIN_ID and LOGIN_PASSWORD environment variables are required")
            sys.exit(1)
        
        self.driver = None
        self.wait = None
        
    def setup_driver(self):
        """Setup Chrome driver with appropriate options"""
        chrome_options = Options()
        
        # Headless mode for GitHub Actions
        if not self.debug_mode:
            chrome_options.add_argument('--headless')
            chrome_options.add_argument('--disable-gpu')
        
        # Common options
        chrome_options.add_argument('--no-sandbox')
        chrome_options.add_argument('--disable-dev-shm-usage')
        chrome_options.add_argument('--disable-blink-features=AutomationControlled')
        chrome_options.add_argument('--window-size=1920,1080')
        chrome_options.add_argument('--start-maximized')
        chrome_options.add_argument('--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36')
        
        # Add experimental options to avoid detection
        chrome_options.add_experimental_option("excludeSwitches", ["enable-automation"])
        chrome_options.add_experimental_option('useAutomationExtension', False)
        
        try:
            # For GitHub Actions, use the system-installed chromedriver
            if os.path.exists('/usr/bin/chromedriver'):
                service = Service('/usr/bin/chromedriver')
                self.driver = webdriver.Chrome(service=service, options=chrome_options)
            else:
                # For local development
                self.driver = webdriver.Chrome(options=chrome_options)
            
            self.driver.implicitly_wait(10)
            self.wait = WebDriverWait(self.driver, 20)
            
            # Execute script to prevent detection
            self.driver.execute_script("Object.defineProperty(navigator, 'webdriver', {get: () => undefined})")
            
            logger.info("Chrome driver initialized successfully")
        except Exception as e:
            logger.error(f"Failed to initialize Chrome driver: {e}")
            sys.exit(1)
    
    def save_debug_info(self, stage):
        """Save debug information if in debug mode"""
        if self.debug_mode:
            try:
                # Take screenshot
                screenshot_path = f'/tmp/{stage}_screenshot.png'
                self.driver.save_screenshot(screenshot_path)
                logger.info(f"Screenshot saved: {screenshot_path}")
                
                # Save page source
                page_source_path = f'/tmp/{stage}_page_source.html'
                with open(page_source_path, 'w', encoding='utf-8') as f:
                    f.write(self.driver.page_source)
                logger.info(f"Page source saved: {page_source_path}")
                
                # Log current URL
                logger.info(f"Current URL: {self.driver.current_url}")
            except Exception as e:
                logger.error(f"Failed to save debug info: {e}")
    
    def login(self):
        """Perform login to GreytHR"""
        try:
            logger.info(f"Navigating to login page: {self.login_url}")
            self.driver.get(self.login_url)
            time.sleep(3)
            
            # Wait for and fill username
            logger.info("Looking for username field...")
            username_field = self.wait.until(
                EC.presence_of_element_located((By.CSS_SELECTOR, 'input[type="text"], input[type="email"], input[name="username"]'))
            )
            username_field.clear()
            username_field.send_keys(self.login_id)
            logger.info("Username entered")
            
            # Fill password
            logger.info("Looking for password field...")
            password_field = self.driver.find_element(By.CSS_SELECTOR, 'input[type="password"]')
            password_field.clear()
            password_field.send_keys(self.login_password)
            logger.info("Password entered")
            
            # Submit form
            logger.info("Submitting login form...")
            try:
                submit_button = self.driver.find_element(By.CSS_SELECTOR, 'button[type="submit"], input[type="submit"]')
                submit_button.click()
            except:
                # Alternative: press Enter on password field
                password_field.submit()
            
            # Wait for login to complete
            time.sleep(5)
            logger.info(f"Login completed, current URL: {self.driver.current_url}")
            
            self.save_debug_info("after_login")
            
            return True
            
        except Exception as e:
            logger.error(f"Login failed: {e}")
            self.save_debug_info("login_error")
            return False
    
    def check_already_signed_in(self):
        """Check if already signed in"""
        try:
            sign_out_buttons = self.driver.find_elements(By.XPATH, "//button[contains(text(), 'Sign Out')]")
            if sign_out_buttons:
                logger.info("✅ Already signed in (Sign Out button found)")
                return True
            return False
        except:
            return False
    
    def handle_sign_in(self):
        """Handle the sign-in process"""
        try:
            # Check if already signed in
            if self.check_already_signed_in():
                return True
            
            # Look for Sign In button
            logger.info("Looking for Sign In button...")
            sign_in_buttons = self.driver.find_elements(By.XPATH, "//button[contains(text(), 'Sign In')]")
            
            if not sign_in_buttons:
                logger.warning("No Sign In button found")
                return False
            
            # Click the first Sign In button
            logger.info(f"Found {len(sign_in_buttons)} Sign In button(s)")
            sign_in_buttons[0].click()
            logger.info("Clicked initial Sign In button")
            
            # Wait for modal to appear
            time.sleep(3)
            self.save_debug_info("after_signin_click")
            
            # Handle modal
            return self.handle_modal()
            
        except Exception as e:
            logger.error(f"Sign-in handling failed: {e}")
            self.save_debug_info("signin_error")
            return False
    
    def handle_modal(self):
        """Handle the work location modal"""
        try:
            logger.info("Checking for work location modal...")
            
            # Method 1: Direct JavaScript execution
            modal_handled = self.driver.execute_script("""
                try {
                    // Check for modal
                    const modals = document.querySelectorAll('gt-popup-modal');
                    console.log('Found ' + modals.length + ' modal(s)');
                    
                    // Look for dropdown button with Office text
                    const dropdownButtons = Array.from(document.querySelectorAll('button')).filter(btn => 
                        btn.textContent.includes('Office') && 
                        btn.offsetParent !== null
                    );
                    
                    if (dropdownButtons.length > 0) {
                        console.log('Clicking dropdown button');
                        dropdownButtons[0].click();
                        return 'dropdown_clicked';
                    }
                    
                    // Alternative: Look for any dropdown button
                    const anyDropdown = document.querySelector('button.dropdown-button');
                    if (anyDropdown) {
                        anyDropdown.click();
                        return 'any_dropdown_clicked';
                    }
                    
                    return 'no_dropdown_found';
                } catch (e) {
                    return 'error: ' + e.message;
                }
            """)
            
            logger.info(f"Dropdown interaction result: {modal_handled}")
            
            if 'clicked' in str(modal_handled):
                time.sleep(2)
                
                # Select Office from dropdown
                office_selected = self.driver.execute_script("""
                    try {
                        // Look for Office option in expanded dropdown
                        const allElements = Array.from(document.querySelectorAll('*'));
                        const officeOptions = allElements.filter(el => 
                            el.textContent.trim() === 'Office' && 
                            el.tagName !== 'BUTTON' &&
                            el.offsetParent !== null
                        );
                        
                        if (officeOptions.length > 0) {
                            officeOptions[0].click();
                            return 'office_selected';
                        }
                        
                        // Alternative: Click div with Office text
                        const divs = document.querySelectorAll('div');
                        for (const div of divs) {
                            if (div.textContent.trim() === 'Office' && !div.querySelector('button')) {
                                div.click();
                                return 'office_div_clicked';
                            }
                        }
                        
                        return 'office_not_found';
                    } catch (e) {
                        return 'error: ' + e.message;
                    }
                """)
                
                logger.info(f"Office selection result: {office_selected}")
                time.sleep(2)
            
            # Click Sign In button in modal
            signin_result = self.driver.execute_script("""
                try {
                    // Find all Sign In buttons
                    const signInButtons = Array.from(document.querySelectorAll('button')).filter(btn => 
                        btn.textContent.trim() === 'Sign In' && 
                        btn.offsetParent !== null
                    );
                    
                    if (signInButtons.length > 1) {
                        // Click the last one (should be in modal)
                        signInButtons[signInButtons.length - 1].click();
                        return 'modal_signin_clicked';
                    } else if (signInButtons.length === 1) {
                        signInButtons[0].click();
                        return 'signin_clicked';
                    }
                    
                    return 'no_signin_button';
                } catch (e) {
                    return 'error: ' + e.message;
                }
            """)
            
            logger.info(f"Sign In click result: {signin_result}")
            
            # Wait for action to complete
            time.sleep(4)
            
            # Verify success
            return self.verify_signin_success()
            
        except Exception as e:
            logger.error(f"Modal handling failed: {e}")
            return False
    
    def verify_signin_success(self):
        """Verify if sign-in was successful"""
        try:
            # Check for Sign Out button
            if self.check_already_signed_in():
                logger.info("✅ Successfully signed in - Sign Out button present")
                return True
            
            # Check for success message
            success_elements = self.driver.find_elements(By.XPATH, "//*[contains(text(), 'Successfully') or contains(text(), 'successfully')]")
            if success_elements:
                logger.info("✅ Successfully signed in - Success message found")
                return True
            
            # Check if modal is gone
            modals = self.driver.find_elements(By.CSS_SELECTOR, 'gt-popup-modal[open]')
            if not modals:
                logger.info("Modal closed - checking final state...")
                if self.check_already_signed_in():
                    return True
                else:
                    logger.warning("⚠️ Modal closed but sign-in status unclear")
                    return False
            else:
                logger.error("❌ Modal still open - sign-in may have failed")
                return False
                
        except Exception as e:
            logger.error(f"Verification failed: {e}")
            return False
    
    def run(self):
        """Main execution flow"""
        try:
            # Setup driver
            self.setup_driver()
            
            # Login
            if not self.login():
                logger.error("Login failed")
                return False
            
            # Handle sign-in
            success = self.handle_sign_in()
            
            if success:
                logger.info("=" * 50)
                logger.info("✅ GREYTHR AUTO SIGN-IN SUCCESSFUL")
                logger.info("=" * 50)
            else:
                logger.error("=" * 50)
                logger.error("❌ GREYTHR AUTO SIGN-IN FAILED")
                logger.error("=" * 50)
            
            self.save_debug_info("final")
            
            return success
            
        except Exception as e:
            logger.error(f"Automation failed: {e}")
            self.save_debug_info("error")
            return False
        finally:
            if self.driver:
                time.sleep(2)
                self.driver.quit()
                logger.info("Driver closed")

def main():
    """Main entry point"""
    automation = GreytHRAutomation()
    success = automation.run()
    
    # Exit with appropriate code
    sys.exit(0 if success else 1)

if __name__ == "__main__":
    main()